File: src/python/oop/ParkingLot/parking/parking_lot.py
from .fee.fee_calculator import FeeCalculator
from .parking_manager import ParkingManager
from .ticket import Ticket
from .vehicle import Vehicle

from datetime import datetime
import uuid

class ParkingLot:
    def __init__(self, parking_manager: ParkingManager, fee_calculator: FeeCalculator):
        self.__parking_manager = parking_manager
        self.__fee_calculator = fee_calculator
        self.__issued_tickets = {}

    def park(self, vehicle: Vehicle) -> Ticket:
        spot = self.__parking_manager.park_vehicle(vehicle)
        if not spot:
            print("Parking spot not found")
            return

        print(f"The parking spot '{spot.number}' assigned")

        ticket = Ticket(uuid.uuid4(), vehicle, spot, datetime.now())
        self.__issued_tickets[ticket.id] = ticket

        print(f"The ticket issued {str(ticket)}")

        # in case of in-memory return ticket copy
        return ticket

    def leave(self, ticket: Ticket):
        if ticket.id not in self.__issued_tickets:
            print(f"Provided ticket not valid")
            return

        original = self.__issued_tickets[ticket.id]
        if original.leave_time:
            print(f"Ticket already processed")
            return

        # ticket validation here
        original.leave_time = datetime.now()
        original.fee = self.__fee_calculator.calculate(ticket)

        print(f"Total fee: {original.fee}")

        return ticket

File: src/go/oop/ParkingLot/internal/parking_lot.go
package internal

import (
	"errors"
	"time"

	"github.com/dobermanch/dev-quests/parking-lot/internal/fees"
	"github.com/dobermanch/dev-quests/parking-lot/internal/models"

	"github.com/google/uuid"
)

type ParkingLot interface {
	Park(models.Vehicle) (models.Ticket, error)
	Leave(models.Ticket) (models.Ticket, error)
}

type ParkingLotImpl struct {
	manager       ParkingManager
	issuedTickets map[string]*models.Ticket
	calculator    fees.FeeCalculator
}

func New() ParkingLotImpl {
	sizes := map[models.SpotSize]int{
		models.CompactSpot:  10,
		models.NormalSpot:   10,
		models.OversizeSpot: 10,
	}

	var lot = ParkingLotImpl{
		manager:       NewManager(sizes),
		issuedTickets: make(map[string]*models.Ticket),
		calculator:    fees.NewCalculator(nil),
	}

	return lot
}

func (pl *ParkingLotImpl) Park(vehicle models.Vehicle) (models.Ticket, error) {
	spot, err := pl.manager.Park(&vehicle)
	if err != nil {
		return models.Ticket{}, err
	}

	if spot == nil {
		return models.Ticket{}, errors.New("the parking spot not found")
	}

	println("The parking spot %s assigned", spot.Number)

	ticket := &models.Ticket{
		Number:      uuid.New().String(),
		Vehicle:     &vehicle,
		Spot:        spot,
		ParkingTime: time.Now(),
	}

	pl.issuedTickets[ticket.Number] = ticket

	println("The ticket has been issued %s", ticket.Number)

	return *ticket, nil
}

func (pl *ParkingLotImpl) Leave(ticket models.Ticket) (models.Ticket, error) {
	if _, ok := pl.issuedTickets[ticket.Number]; !ok {
		return models.Ticket{}, errors.New("provided ticket is not valid")
	}

	originalTicket := pl.issuedTickets[ticket.Number]
	if originalTicket.LeaveTime != nil {
		return models.Ticket{}, errors.New("provided ticket already processed")
	}

	now := time.Now()
	originalTicket.LeaveTime = &now

	fee, err := pl.calculator.Calculate(*originalTicket)
	if err != nil {
		return models.Ticket{}, err
	}

	originalTicket.Fee = fee

	return *originalTicket, nil
}
